src/Lean/Elab/Tactic/BVDecide/Frontend/Normalize/AC.lean

```diff
@@ -1,39 +1,408 @@
+  | none =>
+    modifyGet fun s =>
+      let nextIndex := s.exprToVarIndex.size
+      (nextIndex, { s with
+        exprToVarIndex := s.exprToVarIndex.insert e nextIndex
+        varToExpr := s.varToExpr.push e
+      })
+where
+  /-- Lookup the index of expression `e` in the monadic state. -/
+  getIndex (e : Expr) : VarStateM (Option VarIndex) := do
+    return (← get).exprToVarIndex[e]?
+
+/-- Return the expression that is represented by a specific variable index. -/
+def VarStateM.varToExpr (idx : VarIndex) : VarStateM Expr := do
+  let varToExpr := (← get).varToExpr
+  if h : idx < varToExpr.size then
+    return varToExpr[idx]
+  else
+    throwError "internal error (this is a bug!): index {idx} out of range, \
+      the current state only has {varToExpr.size} variables:\n\n{varToExpr}"
+
+
+/-- Given a binary, associative and commutative operation `op`,
+decompose expression `e` into its variable coefficients.
+
+For example `a ⊕ b ⊕ (a ⊕ c)` will give the coefficients:
+```
+a => 2
+b => 1
+c => 1
+```
+
+Any compound expression which is not an application of the given `op` will be
+abstracted away and treated as a variable (see `VarStateM.exprToVar`).
+
+Note that the output is guaranteed to map at least one variable to a non-zero
+coefficient, *unless* the input expression only contains applications of neutral
+elements (e.g., `0 + (0 + 0)`), in which case the returned coefficients map will
+be empty.
+-/
+def VarStateM.computeCoefficients (op : Op) (e : Expr) : VarStateM CoefficientsMap :=
+  go {} e
+where
+  incrVar (coeff : CoefficientsMap) (e : Expr) : VarStateM CoefficientsMap := do
+    let idx ← exprToVar e
+    return coeff.alter idx (fun c => some <| (c.getD 0) + 1)
+  go (coeff : CoefficientsMap) : Expr → VarStateM CoefficientsMap
+    | e@(AC.bin op' x y) => do
+        if op.isSameKind op' then
+          let coeff ← go coeff x
+          let coeff ← go coeff y
+          return coeff
+        else
+          trace[Meta.AC] "Found binary operation '{op'} {x} {y}', expected '{op}'.\
+            Treating as atom."
+          incrVar coeff e
+    | e => incrVar coeff e
+
+structure SharedCoefficients where
+  common : CoefficientsMap := {}
+  x : CoefficientsMap
+  y : CoefficientsMap
+
+/-- Given two sets of coefficients `x` and `y` (computed with the same variable
+mapping), extract the shared coefficients, such that `x` (resp. `y`) is the sum of
+coefficients in `common` and `x` (resp `y`) of the result.
+
+That is, if `{ common, x', y' } ← SharedCoeffients.compute x y`, then
+  `x[idx] = common[idx] + x'[idx]` and
+  `y[idx] = common[idx] + y'[idx]`
+for all valid variable indices `idx`.
+-/
+def SharedCoefficients.compute (x y : CoefficientsMap) : VarStateM SharedCoefficients := do
+  let swappedXy := x.size > y.size
+  let (x, y) := if swappedXy then (y, x) else (x, y)
+  -- This is O(|x|) = O(min(|x|, |y|)) as we sort by size above.
+  let common : CoefficientsMap := x.fold (init := {}) fun common idx xCnt =>
```

Comment:
Probably no need for type annotations here right?

Reply:
No *need*, but I prefer to keep it for readability.